File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第4章：对比度增强/4.2/line.cpp
#include<opencv2/core/core.hpp>
#include<opencv2/highgui/highgui.hpp>
#include<opencv2/imgproc/imgproc.hpp>
#include<string>
#include<iostream>
cv::Mat image;//输入图像
cv::Mat contrastImage;//输出图像
//图像的宽高
int height;
int width;
int value = 40;
float segValue = float(value);
const int MAX_VALUE = 120;
std::string contrastWindow = "对比度";
void callback_value(int, void*)
{
	//斜率 a
	float a = float(value) / segValue;
	//调整对比度
	for (int r = 0; r < height; r++)
	{
		for (int c = 0; c < width; c++)
		{
			float pixel = a*image.at<uchar>(r, c);
			pixel = round(pixel);
			//截断
			if (pixel > 255)
				pixel = 255;
			contrastImage.at<uchar>(r, c) = uchar(pixel);
		}
	}
	//显示调整对比度后的效果
	cv::imshow(contrastWindow, contrastImage);
}
int main(int argc, char*argv[])
{
	//输入图片
	image = cv::imread(argv[1], CV_LOAD_IMAGE_GRAYSCALE);
	if (!image.data)
	{
		std::cout << "没有图片" << std::endl;
		return -1;
	}
	//显示原图
	cv::imshow("原图", image);
	//图像的宽高
	height = image.rows;
	width = image.cols;
	contrastImage = cv::Mat::zeros(image.size(), CV_8UC1);
	cv::namedWindow(contrastWindow, CV_WINDOW_AUTOSIZE);
	cv::createTrackbar("value", contrastWindow, &value, MAX_VALUE, callback_value);
	callback_value(0, 0);
	cv::waitKey(0);
	return 0;
}

File: OpenCV演算法精解：基於Python與C++/code/unzip/OpenCV算法精解-代码/第4章：对比度增强/4.2/line.py
# -*- coding: utf-8 -*-
import cv2
import numpy as np
import sys
#主函数
if __name__ =="__main__":
    if len(sys.argv) > 1:
        image = cv2.imread(sys.argv[1],cv2.IMREAD_GRAYSCALE)
    else:
        print "Usge:python lineContrast.py imageFile"
    MAX_VALUE = 120
    value = 120
    #调整对比度后，图像的效果显示窗口
    cv2.namedWindow("contrast",cv2.WND_PROP_AUTOSIZE)
    #调整系数，观察图像的变化
    def callback_contrast(_value):
        #通过线性运算，调整图像对比度
        a = float(_value)/40.0
        contrastImage = a*image
        contrastImage[contrastImage>255]=255
        contrastImage = np.round(contrastImage)
        contrastImage = contrastImage.astype(np.uint8)
        cv2.imshow("contrast",contrastImage)
        cv2.imwrite("contrast.jpg",contrastImage)
    callback_contrast(value)
    cv2.createTrackbar("value","contrast",value,MAX_VALUE,callback_contrast)
    cv2.waitKey(0)
    cv2.destroyAllWindows()
